src/utils/changelog.ts

```diff
@@ -0,0 +1,396 @@
+    );
+    changelogSections.push(milestone.description);
+    changelogSections.push(
+      `PRs: ${milestones[milestoneNum].prs.map(pr => `#${pr}`).join(', ')}`
+    );
+  }
+
+  if (leftovers.length > 0) {
+    changelogSections.push(`## Various fixes & improvements`);
+    changelogSections.push(leftovers.map(formatCommit).join('\n'));
+  }
+
+  return changelogSections.join('\n\n');
+}
+
+interface CommitInfoResult {
+  repository: {
+    [hash: string]: {
+      associatedPullRequests: {
+        nodes: Array<{
+          number: string;
+          milestone: {
+            number: string;
+          };
+        }>;
+      };
+    } | null;
+  };
+}
+
+async function getMilestoneAndPRFromCommits(
+  hashes: string[]
+): Promise<
+  Record</* hash */ string, { pr: string | null; milestone: string | null }>
+> {
+  if (hashes.length === 0) {
+    return {};
+  }
+
+  const commitsQuery = hashes
+    // We need to prefix the hash value (with `C` here) when using it as an
+    // alias as aliases cannot start with a number but hashes can.
+    .map(hash => `C${hash}: object(oid: "${hash}") {...PRFragment}`)
+    .join('\n');
+
+  const { repo, owner } = await getGlobalGithubConfig();
+  const commitInfo = ((await getGitHubClient().graphql(`{
+    repository(name: "${repo}", owner: "${owner}") {
+      ${commitsQuery}
+    }
+  }
+
+  fragment PRFragment on Commit {
+    associatedPullRequests(first: 1) {
+      nodes {
+        number
+        milestone {
+          number
+        }
+      }
+    }
+  }`)) as CommitInfoResult).repository;
+
+  return Object.fromEntries(
+    Object.entries(commitInfo).map(([hash, commit]) => [
+      // Strip the prefix on the hash we used to workaround in GraphQL
+      hash.slice(1),
+      {
+        pr: commit?.associatedPullRequests.nodes[0]?.number ?? null,
+        milestone:
+          commit?.associatedPullRequests.nodes[0]?.milestone?.number ?? null,
+      },
+    ])
+  );
+}
+
+interface MilestonesDetailsResult {
+  repository: {
+    [number: string]: {
+      title: string;
+      description: string;
+      state: 'OPEN' | 'CLOSED';
+    };
+  };
+}
+
+async function getMilestonesDetails(
+  milestones: string[]
+): Promise<Record<string, Milestone>> {
+  if (milestones.length === 0) {
+    return {};
+  }
+
+  const milestoneQuery = milestones
+    .map(
+      number =>
+        // We need to prefix the milestone number (with `M` here) when using it
+        // as an alias as aliases cannot start with a number.
+        `M${number}: milestone(number: ${number}) {...MilestoneFragment}`
+    )
+    .join('\n');
+
+  const { repo, owner } = await getGlobalGithubConfig();
+  const milestoneInfo = ((await getGitHubClient().graphql(`{
+    repository(name: "${repo}", owner: "${owner}") {
+      ${milestoneQuery}
+    }
+  }
+
+  fragment MilestoneFragment on Milestone {
+    title
+    description
+    state
+  }
+`)) as MilestonesDetailsResult).repository;
```

Comment:
Maybe worth exploring [a query builder](https://www.npmjs.com/package/gql-query-builder) in the future if this becomes something we do more of?

Reply:
Yup, totally agree with that. And a validator for tests. Overkill for now though.